ner_v2/detectors/property_assignor.py

```diff
@@ -0,0 +1,334 @@
+        it is returned as is.
+
+        Args:
+            text (str/unicode): text to clean up
+            ignore_puncts (bool): whether to drop punctuations from the text and replace them with a space
+
+        Returns:
+            unicode or Any: if text is non empty string, a decoded string with optionally punctuations replaced.
+                Otherwise the text is returned as is.
+        """
+        if text and isinstance(text, six.string_types):
+            text = v2_utils.ensure_str(text)
+            if ignore_puncts:
+                text = replace_puncts(text=text, replace_with=u" ", except_puncts=(u"_",))
+        return text
+
+    def assign_properties(self, text, detected_spans, bot_message=None, ignore_puncts=True):
+        # type: (unicode, List[unicode], Optional[unicode], bool) -> List[AssignedProperty]
+        """
+        Assign properties to `detected_spans` based on their contexts
+
+        Rough pseudocode:
+            - First drop punctuations from text and bot_message if ignore_puncts is True
+            - Split both by whitespace and join them back by single space
+            - Replace all detected spans in text with an entity placeholder
+            - For each replaced entity placeholder
+                For each property
+                    Match prefix within N words before this entity
+                    Match suffix within M words after this entity
+                    If both prefix and suffix must match
+                        If both matches are found
+                            Assign this property
+                    Else if prefix is matched
+                        Assign this property
+                    Else if suffix is matched
+                        Assign this property
+                    Else if bot message is given and has required terms
+                        Assign this property
+
+                    If property is assigned
+                        Break
+
+        Args:
+            text (str/unicode): the text from which entities were detetced
+            detected_spans (list of str/unicode): list of substrings of text that were detected as some entity
+            bot_message (str/unicode, optional): message from the bot in the last turn in the dialogue.
+                Defaults to None
+            ignore_puncts (bool, optional): whether to drop punctuations from the text and bot message
+                and replace them with a space. Defaults to True
+
+        Returns:
+            list: list of AssignedProperty (namedtuple) of same length as `detected_spans`.
+                Each AssignedProperty has two members 
+                    `property` which can be None or str - the property assigned to ith span and
+                    `reason` which can be None or str - why was this property assign. Can be one of 
+                        "prefix_suffix" - both prefix and suffix patterns matched
```

Comment:
Actual string which led to detection of prefix / suffix, etc. is not being returned or logged. Might become hard to debug if wrong property is being detected. If possible return that as well.

Reply:
To be included in final return type